cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java

```diff
@@ -0,0 +1,98 @@
+import io.cdap.cdap.api.service.worker.RemoteTaskException;
+import io.cdap.cdap.api.service.worker.RunnableTaskRequest;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.internal.remote.RemoteClientFactory;
+import io.cdap.cdap.common.internal.remote.RemoteTaskExecutor;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.AuthenticationConfigException;
+import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.worker.PushAppTask;
+import io.cdap.common.http.HttpRequestConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import javax.inject.Inject;
+
+/**
+ * Remote implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside task worker.
+ */
+public class RemoteSourceControlOperationRunner implements SourceControlOperationRunner {
+
+  private static final Gson GSON = new GsonBuilder().create();
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteSourceControlOperationRunner.class);
+  private final RemoteTaskExecutor remoteTaskExecutor;
+
+  @Inject
+  RemoteSourceControlOperationRunner(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+                                     RemoteClientFactory remoteClientFactory) {
+    int readTimeout = cConf.getInt(Constants.TaskWorker.SOURCE_CONTROL_HTTP_CLIENT_READ_TIMEOUT_MS);
+    int connectTimeout = cConf.getInt(Constants.TaskWorker.SOURCE_CONTROL_HTTP_CLIENT_CONNECTION_TIMEOUT_MS);
+    HttpRequestConfig httpRequestConfig = new HttpRequestConfig(connectTimeout, readTimeout, false);
+    this.remoteTaskExecutor = new RemoteTaskExecutor(cConf, metricsCollectionService, remoteClientFactory,
+                                                     RemoteTaskExecutor.Type.TASK_WORKER, httpRequestConfig);
+  }
+
+  @Override
+  public PushAppResponse push(PushAppContext pushAppContext)
+    throws PushFailureException, NoChangesToPushException, AuthenticationConfigException {
+    try {
+      RunnableTaskRequest request =
+        RunnableTaskRequest.getBuilder(PushAppTask.class.getName()).withParam(GSON.toJson(pushAppContext)).build();
+      
+      LOG.debug("Pushing application {} to linked repository", pushAppContext.getAppToPush());
+      byte[] result = remoteTaskExecutor.runTask(request);
+      return GSON.fromJson(new String(result, StandardCharsets.UTF_8), PushAppResponse.class);
+    } catch (RemoteExecutionException e) {
+      RemoteTaskException remoteTaskException = e.getCause();
+      String exceptionClass = remoteTaskException.getRemoteExceptionClassName();
+      String exceptionMessage = remoteTaskException.getMessage();
+
+      if (NoChangesToPushException.class.getName().equals(exceptionClass)) {
```

Comment:
I am wondering if we should just improve io.cdap.cdap.api.service.worker.RemoteExecutionException#fromBasicThrowable to instantiate the remove class. The way we do here looses any root causes

Reply:
I think the problem here is that the exception classes in the cause are not the same because of how RemoteExecutionException is implemented, plus everything that calls runTask() needs to do similar logic to propagate the exception. +1 to improving it, though it should be in a separate PR. Can you open an enhancement jira and add a todo?

Reply:
what about the other exceptions like AuthenticationConfigException, PushFailureException, and IOException?

Reply:
Added TODO ticket: https://cdap.atlassian.net/browse/CDAP-20418

`AuthenticationConfigException` is handled in below if section. `PushFailureException`, and `IOException` are both transformed to `PushFailureException`

Reply:
shouldn't it do the same thing and try to preserve the IOException? Otherwise the remote and local implementations will throw different exceptions in different situations.

If there isn't really a meaningful way to handle IOException, we should wrap it into a PushFailureException in the other implementation as well so that they are consistent.

Reply:
@albertshau We don't throw `IOException` in the local implementation either (it was wrapped into a `PushFailureException`. This `IOException` in remote implementation was thrown by the `PushAppTask#context.writeResult`.

Reply:
Offline discusses. IOException is no need to preserve